<br>

<img src="../image/Logo/logo_elia_group.png" width = 200>

<br>

# Answers for Pandas Chapter 41-48

In the following you will find all the answers to the exercises of the Python Training for Beginners by the Elio Group. Have fun! 

## (41) Pandas - Data Exploration

In [3]:
import pandas as pd
pv_power = pd.read_csv("./data/energy/ods032.csv", sep = ';')

### Exercise

1. Select the first 7 rows, and first 2 columns of `pv_power`, positionally. 
2. Select rows 10, 15, ans 30, and the columns `Resolution code` and `Region`.

&#9200; Still time left? Try this one:

3. Select the first 5 rows and columns `Datetime` and `Monitored capacity` of the pv_power table. Hint: you'll need to string together 2 different types of selectors. And remember, google is your friend.

### Solutions

In [4]:
#1. Task 

pv_power.iloc[:7, :2]

Datetime Resolution code
0  2022-06-21T23:45:00+02:00           PT15M
1  2022-06-21T23:45:00+02:00           PT15M
2  2022-06-21T23:45:00+02:00           PT15M
3  2022-06-21T23:45:00+02:00           PT15M
4  2022-06-21T23:45:00+02:00           PT15M
5  2022-06-21T23:45:00+02:00           PT15M
6  2022-06-21T23:45:00+02:00           PT15M

In [5]:
# 2. Task

pv_power.loc[[10, 15, 30], ["Resolution code", "Region"]]

Resolution code         Region
10           PT15M        Antwerp
15           PT15M        Antwerp
30           PT15M  East-Flanders

In [6]:
# 3. Task 

pv_power.iloc[:5].loc[:, ["Datetime", "Monitored capacity"]]

Datetime  Monitored capacity
0  2022-06-21T23:45:00+02:00             634.933
1  2022-06-21T23:45:00+02:00             431.127
2  2022-06-21T23:45:00+02:00             514.584
3  2022-06-21T23:45:00+02:00            1127.039
4  2022-06-21T23:45:00+02:00             195.154

## (43) Pandas - Data Joining 

In [2]:
import pandas as pd

### Exercise

Now it's your turn! 

- Check out the following two data frames:

In [12]:
df1_job = pd.DataFrame({'employee': ['Bob', 'Jake', 'Ahmed', 'Sue'],
                    'group': ['Data Science', 'Data Engineering', 'Data Engineering', 'Data Analyst']})
df2_hired = pd.DataFrame({'employee': ['Ahmed', 'Bob', 'Jake', 'Sue'],
                    'hire_date': [2006, 2008, 2014, 2021]})

- print out the first three rows of each DataFrame before merging
- Merge the following dataframes based on their name and save it into a new dataframe called df_job_hired
- print out the first 3 rows

In [13]:
df1_job.head(n=3)

employee             group
0      Bob      Data Science
1     Jake  Data Engineering
2    Ahmed  Data Engineering

In [14]:
df2_hired.head(n=3)

employee  hire_date
0    Ahmed       2006
1      Bob       2008
2     Jake       2014

In [16]:
df_job_hired = df1_job.merge(df2_hired, on = "employee")
df_job_hired

employee             group  hire_date
0      Bob      Data Science       2008
1     Jake  Data Engineering       2014
2    Ahmed  Data Engineering       2006
3      Sue      Data Analyst       2021

### Advanced Exercise

In the DataFrame `df2_hired` somebody changed the name of the employee column. In addition, a column called `responsibility` was added. 

- See if both `responsibility` columns represent the same values and can be merged into one column.

- Merge the two DataFrames, save them in a new variable called `df_job_hired_2`.
- Add suffixes ("_job", "_hired") so you can still distinguish the columns after merging. 
- In the end, print out the DataFrame.

&#128161; <ins>Hint</ins>: Use the parameters left_on, right_on of the .merge() function. And don't forget, you can always check the documentation if needed.

In [2]:
import pandas as pd

In [3]:
df1_job = pd.DataFrame({'employee': ['Bob', 'Jake', 'Lisa', 'Sue'],
                       'group': ['Data Science', 'Data Engineering', 'Data Engineering', 'Data Analyst'],
                    'responsibility': ["Energy Forecast", "Data pipelines", "Data pipelines", "Interpreting Results"]})

df2_hired = pd.DataFrame({'employee_names': ['Lisa', 'Bob', 'Jake', 'Sue'],
                         'hire_date': [2004, 2008, 2012, 2014],
                         'responsibility': ["yes", "no", "no", "yes"]})

In [5]:
df1_job

employee             group        responsibility
0      Bob      Data Science       Energy Forecast
1     Jake  Data Engineering        Data pipelines
2     Lisa  Data Engineering        Data pipelines
3      Sue      Data Analyst  Interpreting Results

In [6]:
df2_hired

employee_names  hire_date responsibility
0           Lisa       2004            yes
1            Bob       2008             no
2           Jake       2012             no
3            Sue       2014            yes

In [7]:
df1_job.responsibility == df2_hired.responsibility

0    False
1    False
2    False
3    False
Name: responsibility, dtype: bool

In [8]:
df_job_hired_2 = df1_job.merge(df2_hired, left_on = "employee", right_on = "employee_names", suffixes = ("_job", "_hired"))

In [17]:
print(df_job_hired_2)

  employee             group    responsibility_job employee_names  hire_date  \
0      Bob      Data Science       Energy Forecast            Bob       2008   
1     Jake  Data Engineering        Data pipelines           Jake       2012   
2     Lisa  Data Engineering        Data pipelines           Lisa       2004   
3      Sue      Data Analyst  Interpreting Results            Sue       2014   

  responsibility_hired  
0                   no  
1                   no  
2                  yes  
3                  yes  


### Exercise

In the following, two dataframes (`dp1_df`, and `dp2_df`) are given. They both describe a number of employees working on a data related project together. They share the same column names. 

- Concatenate the dataframes and save them in a new dataframe called `data_project_df`.
- When you concatenated the dataframes, have a look at the index - it is not continuous. 
- Add the parameter "ignore_index = True" to create a continuous index.

In [4]:
dp1_df = pd.DataFrame([{'Name': 'Kelly', 'Role': 'Team Lead ML', 'Location': 'State Street'},
                         {'Name': 'Sally', 'Role': 'Data Scientist', 'Location': 'Washington Avenue'},
                         {'Name': 'James', 'Role': 'Data Engineer', 'Location': 'Washington Avenue'}])
dp2_df = pd.DataFrame([{'Name': 'James', 'Role': 'Analyst', 'Location': '1024 Billiard Avenue'},
                           {'Name': 'Mike', 'Role': 'Regulations', 'Location': 'Fraternity House #22'},
                           {'Name': 'Sally', 'Role': 'MlOps', 'Location': '512 Wilson Crescent'}])

In [24]:
data_project_df = pd.concat([dp1_df, dp2_df], ignore_index = True, verify_integrity = True )

In [ ]:
data_project_df

## (45) Pandas - Conditional Selections

### Excercise

1. Look at the dataframe `energy_flow` above and select the control area "Germany". 
2. Safe your selection into a variable calles `germany`.
3. Look at the first three rows to see if it worked.

In [23]:
energy_flow = pd.read_csv("../data/energy/physical_flow_2021_1_01.csv", sep = ";", index_col = 0)

In [24]:
energy_flow.head()

Resolution code   Control area  Physical Flow Value
Datetime                                                                     
2021-12-01T23:45:00+01:00           PT15M    Netherlands             -419.704
2021-12-01T23:45:00+01:00           PT15M  UnitedKingdom             1021.774
2021-12-01T23:45:00+01:00           PT15M     Luxembourg              -53.708
2021-12-01T23:45:00+01:00           PT15M        Germany            -1001.788
2021-12-01T23:45:00+01:00           PT15M         France              581.908

In [ ]:
germany = energy_flow[energy_flow["Control area"] == "Germany"]

In [ ]:
germany.head(n=3)

<br>

### Exercise

Now it's your turn. 

1. Update the following cell to upload the data with 
    - Datetime being a **DatetimeIndex** 
    - this DatetimeIndex being the only Index

In [2]:
import pandas as pd

In [4]:
physical_flow = pd.read_csv("../data/energy/physical_flow_2021_1_01.csv", sep = ";", parse_dates = True, index_col = 0)

In [10]:
physical_flow.head()

Resolution code   Control area  Physical Flow Value
Datetime                                                                     
2021-12-01 23:45:00+01:00           PT15M    Netherlands             -419.704
2021-12-01 23:45:00+01:00           PT15M  UnitedKingdom             1021.774
2021-12-01 23:45:00+01:00           PT15M     Luxembourg              -53.708
2021-12-01 23:45:00+01:00           PT15M        Germany            -1001.788
2021-12-01 23:45:00+01:00           PT15M         France              581.908

2. Resemple your data to get the mean() by hour

In [18]:
hourly_mean = physical_flow.resample("H").mean()

In [20]:
hourly_mean.columns = ["Physical_Flow_mean"]

In [22]:
hourly_mean.head(n=3)

Physical_Flow_mean
Datetime                                     
2021-12-01 00:00:00+01:00           264.94165
2021-12-01 01:00:00+01:00           312.89260
2021-12-01 02:00:00+01:00           358.70945

3. Remove the timezone from `physical_flow` and store the data in a variable called `none_tz`

In [14]:
none_tz = physical_flow.tz_localize(None).head()

In [15]:
none_tz.tail(3)

Resolution code Control area  Physical Flow Value
Datetime                                                             
2021-12-01 23:45:00           PT15M   Luxembourg              -53.708
2021-12-01 23:45:00           PT15M      Germany            -1001.788
2021-12-01 23:45:00           PT15M       France              581.908

4. Select all data from `none_tz` that is from the 2021-12-01 and between 5am and 5:45am, store the data in a variable called `early_morning`

In [16]:
late_night = none_tz.loc["2021-12-01 23:00:00":"2021-12-01 23:45:00"]

In [23]:
late_night

Resolution code   Control area  Physical Flow Value
Datetime                                                               
2021-12-01 23:45:00           PT15M    Netherlands             -419.704
2021-12-01 23:45:00           PT15M  UnitedKingdom             1021.774
2021-12-01 23:45:00           PT15M     Luxembourg              -53.708
2021-12-01 23:45:00           PT15M        Germany            -1001.788
2021-12-01 23:45:00           PT15M         France              581.908

### Exercise 

Let's play around a bit using the physical flow data again.

In [31]:
import pandas as pd

In [32]:
physical_flow = pd.read_csv("../data/energy/physical_flow_2021_1_01.csv", sep = ";")

In [33]:
physical_flow.head(n=3)

Datetime Resolution code   Control area  \
0  2021-12-01T23:45:00+01:00           PT15M    Netherlands   
1  2021-12-01T23:45:00+01:00           PT15M  UnitedKingdom   
2  2021-12-01T23:45:00+01:00           PT15M     Luxembourg   

   Physical Flow Value  
0             -419.704  
1             1021.774  
2              -53.708

1. Convert your `Datetime` data into real Datetime objects by using `pd.to_datetime`

In [34]:
physical_flow["Datetime"] = pd.to_datetime(physical_flow["Datetime"], format = "%Y-%m-%dT%H:%M:%S%z")

2. Print out the name of the day from 4th row

In [35]:
day_name = physical_flow.loc[4, "Datetime"].day_name()

In [37]:
print(day_name)

Wednesday


3. Let's now say, that there was a huge imbalance on the net on the 23th of July in 2002.
    - Turn the following string into a datetime using `pd.to_datetime`
    - print out a String saying "Exact time of imbalance: Tuesday, Jul 07th 23:07:44:000000"

In [42]:
imbalance_date = "23/07/2002 23:07:44"
imbalance_date = pd.to_datetime(imbalance_date)
imbalance_date

Timestamp('2002-07-23 23:07:44')

In [43]:
print(imbalance_date.strftime("Exact time of imalance: %A, %b %mth %H:%M:%S:%f"))

Exact time of imalance: Tuesday, Jul 07th 23:07:44:000000


## (47) Pandas - Advanced Manipulations

### Exercise

Let's do another one. <br>
1. First, create a new function called `flow_check` that returns ">200 MW" if the `Physical Flow` value is bigger than 200 MW and "<200 MW" if it is less than 200 MW.
2. Create a new column called `Flow Size` that applies this function onto your dataframe

In [4]:
import pandas as pd

In [6]:
physical_flow = pd.read_csv("../data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")

In [7]:
ph_flow_short = physical_flow.loc[:10,"Resolution code":"Loading"]
ph_flow_short

Resolution code Asset Elia ID                    Asset name     Asset type  \
0            PT15M  380_51_LILLO          DOEL-MERCATOR 380.51  Internal Line   
1            PT15M  380_74_HORTA         HORTA-MERCATOR 380.74  Internal Line   
2            PT15M        380_29      RILLAND-ZANDVLIET 380.29        Tieline   
3            PT15M       380_111        GEZELLE-STEVIN 380.111  Internal Line   
4            PT15M        380_79       AVELGEM-MASTAING 380.79        Tieline   
5            PT15M  380_33_BRUEG     BRUEGEL-COURCELLES 380.33  Internal Line   
6            PT15M        380_26         DOEL-ZANDVLIET 380.26  Internal Line   
7            PT15M       380_109  GEZELLE-VAN MAERLANT 380.109  Internal Line   
8            PT15M       380_107  GEZELLE-VAN MAERLANT 380.107  Internal Line   
9            PT15M        380_10          ACHENE-GRAMME 380.10  Internal Line   
10           PT15M        380_28    MAASBRACHT-VAN EYCK 380.28        Tieline   

    Base Voltage AC/DC  Physical Flow  Loading  
0          380.0    AC      13.754076     0.68  
1          380.0    AC     238.395599     7.01  
2          380.0    AC     232.342749    10.78  
3          380.0    AC     121.828900     3.72  
4          380.0    AC       0.015625     0.00  
5          380.0    AC      24.655853     1.77  
6          380.0    AC     298.491598    18.90  
7          380.0    AC       0.030151     0.00  
8          380.0    AC       0.030151     0.00  
9          380.0    AC     440.964253    24.38  
10         380.0    AC      57.433527     3.64

In [8]:
def flow_size(row):
    if row["Physical Flow"] >= 200:
        return ">200 MW"
    else: 
        return "<200 MW"

In [10]:
ph_flow_short.loc[:, "Flow Size"] = ph_flow_short.apply(flow_size, axis = 1)
ph_flow_short.tail()

Resolution code Asset Elia ID                    Asset name     Asset type  \
6            PT15M        380_26         DOEL-ZANDVLIET 380.26  Internal Line   
7            PT15M       380_109  GEZELLE-VAN MAERLANT 380.109  Internal Line   
8            PT15M       380_107  GEZELLE-VAN MAERLANT 380.107  Internal Line   
9            PT15M        380_10          ACHENE-GRAMME 380.10  Internal Line   
10           PT15M        380_28    MAASBRACHT-VAN EYCK 380.28        Tieline   

    Base Voltage AC/DC  Physical Flow  Loading Flow Size  
6          380.0    AC     298.491598    18.90   >200 MW  
7          380.0    AC       0.030151     0.00   <200 MW  
8          380.0    AC       0.030151     0.00   <200 MW  
9          380.0    AC     440.964253    24.38   >200 MW  
10         380.0    AC      57.433527     3.64   <200 MW

### Advanced Exercise
In your dataframe, the physical flow indicates how heavily the line is loaded relative to the maximum possible line loading.

1. Create a new column that returns the maximum possible loading.
2. For instance, if the physical flow is 13.75 MW and that equals 0.68% loading, then the max. loading is 20,22 MW. 
3. Name the new column "Max. Loading"

In [13]:
import pandas as pd

In [14]:
physical_flow = pd.read_csv("../data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")

In [16]:
ph_flow_short = physical_flow.loc[:10,"Resolution code":"Loading"]
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  
0         380.0    AC      13.754076     0.68  
1         380.0    AC     238.395599     7.01  
2         380.0    AC     232.342749    10.78  
3         380.0    AC     121.828900     3.72  
4         380.0    AC       0.015625     0.00

In [29]:
def max_calculator(row):
    if (row["Physical Flow"] > 0) & (row["Loading"] > 0):
        return row["Physical Flow"] / row["Loading"] *100
    else:
        return 0

In [30]:
ph_flow_short.loc[:, "Max. Loading"] = ph_flow_short.apply(max_calculator, axis = 1)
ph_flow_short.head()

Resolution code Asset Elia ID                Asset name     Asset type  \
0           PT15M  380_51_LILLO      DOEL-MERCATOR 380.51  Internal Line   
1           PT15M  380_74_HORTA     HORTA-MERCATOR 380.74  Internal Line   
2           PT15M        380_29  RILLAND-ZANDVLIET 380.29        Tieline   
3           PT15M       380_111    GEZELLE-STEVIN 380.111  Internal Line   
4           PT15M        380_79   AVELGEM-MASTAING 380.79        Tieline   

   Base Voltage AC/DC  Physical Flow  Loading  Max. Loading  
0         380.0    AC      13.754076     0.68   2022.658205  
1         380.0    AC     238.395599     7.01   3400.793142  
2         380.0    AC     232.342749    10.78   2155.313073  
3         380.0    AC     121.828900     3.72   3274.970430  
4         380.0    AC       0.015625     0.00      0.000000

### Exercise 

This exercise is all about vectorization: 

- Check out the following dataframe
- Create a new column "Loading_grouped" in the DataFrame `flow_info` that groups the Loding (in %) using vectorization
- Loading less then 5 % is categorized as "small", less then 20 % as "medium" and more then 20 % as "large"
- print out the the last 10 rows of the updates dataframe

In [20]:
import numpy as np 

In [22]:
physical_flow = pd.read_csv("../data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")

flow_info = physical_flow.loc[:, "Asset name": "Loading"]
flow_info.head(3)

Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   

   Loading  
0     0.68  
1     7.01  
2    10.78

In [23]:
conditions = [
    flow_info["Loading"] < 5, # first condition to test: if true return choice1, if false check next condition
    flow_info["Loading"] <= 20, # second condition to test: if true return choice2, if false check next condition
    flow_info["Loading"] > 20 # third condition to test: if true return choice3, if false default value is returned
]

choices = [
    "small",  # choice1
    "medium", # choice2
    "large",   # choice3
]

flow_info.loc[:,"Loading_grouped"] = np.select(conditions, choices) #  default value is the value if non of the conditions are true

In [25]:
print(flow_info.Loading_grouped.value_counts())
flow_info.tail(10)

medium    2170
large     1658
small     1248
Name: Loading_grouped, dtype: int64


Asset name     Asset type  Base Voltage AC/DC  \
5066  MAASBRACHT-VAN EYCK 380.28        Tieline         380.0    AC   
5067         GRAMME-LIXHE 380.11  Internal Line         380.0    AC   
5068        DOEL-MERCATOR 380.52  Internal Line         380.0    AC   
5069  HORTA-VAN MAERLANT 380.104  Internal Line         380.0    AC   
5070        DOEL-MERCATOR 380.51  Internal Line         380.0    AC   
5071    COURCELLES-GRAMME 380.31  Internal Line         380.0    AC   
5072       HORTA-MERCATOR 380.74  Internal Line         380.0    AC   
5073     BRUEGEL-MERCATOR 380.36  Internal Line         380.0    AC   
5074  HORTA-VAN MAERLANT 380.103  Internal Line         380.0    AC   
5075   BRUEGEL-COURCELLES 380.34  Internal Line         380.0    AC   

      Physical Flow  Loading Loading_grouped  
5066      64.885999     4.11           small  
5067     254.106699    14.87          medium  
5068     550.735256    37.37           large  
5069     248.927759     7.20          medium  
5070       2.042895     0.10           small  
5071     154.208288     8.47          medium  
5072      95.991292     3.32           small  
5073     748.364497    47.38           large  
5074     207.980390     6.19          medium  
5075      42.436825     2.69           small

### Advanced Exercise 

In the follwing, you will calculate the maximum load again. Similar to the exercise from above. However, this time you will use vectorization again.

- Work with the dataframe `flow_info`
- Create a new column called "Max_load" that shows the maximum possible load
        - For instance, if the physical flow is 13.75 MW and that equals 0.68% loading, then the max. loading is 20,22 MW
- Hint: Pay attention to zero values

In [ ]:
import numpy as np 
physical_flow = pd.read_csv("../data/energy/physical_flow_high_voltage_2022_may_30.csv", sep = ";")

flow_info = physical_flow.loc[:, "Asset name": "Loading"]
flow_info.head(3)

In [31]:
flow_info.loc[:, "Max_load"] = np.where(
    (flow_info["Physical Flow"] > 0) & (flow_info["Loading"] > 0), # <-- condition
    flow_info["Physical Flow"] / flow_info["Loading"] * 100, # <-- return if true
    0 #<-- return if false
    )
flow_info.head(3)

Asset name     Asset type  Base Voltage AC/DC  Physical Flow  \
0      DOEL-MERCATOR 380.51  Internal Line         380.0    AC      13.754076   
1     HORTA-MERCATOR 380.74  Internal Line         380.0    AC     238.395599   
2  RILLAND-ZANDVLIET 380.29        Tieline         380.0    AC     232.342749   

   Loading Loading_grouped     Max_load  
0     0.68           small  2022.658205  
1     7.01          medium  3400.793142  
2    10.78          medium  2155.313073

## (48) Pandas - Data Visualization


### Exercise

In [ ]:
plt.axhline(y=0.5, color='r', linestyle='--')